exercises/5.1/template_ws/ros2/lesson_perception/src/perception_node.cpp

```diff
@@ -0,0 +1,211 @@
+#include "rclcpp/rclcpp.hpp"
+#include <tf2/transform_datatypes.h>
+#include <tf2_ros/transform_listener.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <sensor_msgs/msg/point_cloud2.hpp>
+#include <rclcpp/qos.hpp>
+
+// PCL specific includes
+#include <pcl_conversions/pcl_conversions.h>
+#include "pcl_ros/transforms.hpp"
+
+#include <pcl/filters/voxel_grid.h>
+#include <pcl/filters/passthrough.h>
+#include <pcl/sample_consensus/method_types.h>
+#include <pcl/sample_consensus/model_types.h>
+#include <pcl/segmentation/sac_segmentation.h>
+#include <pcl/filters/extract_indices.h>
+#include <pcl/segmentation/extract_clusters.h>
+#include <pcl/filters/crop_box.h>
+#include <pcl/filters/statistical_outlier_removal.h>
+#include <tf2/convert.h>
+#include <tf2_eigen/tf2_eigen.h>
+#include "tf2/LinearMath/Quaternion.h"
+#include <pcl/segmentation/extract_polygonal_prism_data.h>
+
+class PerceptionNode : public rclcpp::Node
+{
+    public:
+        PerceptionNode()
+        : Node("perception_node", rclcpp::NodeOptions()
+                                          .allow_undeclared_parameters(true)
+                                          .automatically_declare_parameters_from_overrides(true))
```

Comment:
You can also use this complementary function for declaring & getting parameters, if you prefer.  Throwing it at the top of the file will mean you can single-line declare and retrieve a parameter or its default later.  Only catch is making sure you pass default strings as `std::string("thing here")`, as it otherwise has trouble.  I use this function all the time; on the other hand it might be nice to familiarize them with 'vanilla' ROS stuff without external convenience functions.
```
template <class T>
void param(rclcpp::Node::SharedPtr node, const std::string& name, T& output, const T& default_value)
{
  output = default_value;
  node->declare_parameter(name, default_value);
  node->get_parameter(name, output);
  return;
}
```